rest/oidc_api_test.go

```diff
@@ -1,139 +1,865 @@
+					UserPrefix:         "foo",
+					ValidationKey:      base.StringPointer("qux"),
+					Register:           true,
+					DiscoveryURI:       issuerFoo + auth.OIDCDiscoveryConfigPath,
+					IncludeAccessToken: true,
+				},
+			},
+			defaultProvider:          "foo",
+			wantUsername:             "foo_noah",
+			authURL:                  "/db/_oidc?provider=foo&offline=true",
+			register:                 true,
+			includeAccessToken:       true,
+			wantRefreshTokenExgError: true,
+		}, {
+			name: "token refresh no ID token received from OAuth server",
+			providers: auth.OIDCProviderMap{
+				"foo": &auth.OIDCProvider{
+					Name:               "foo",
+					Issuer:             issuerFoo,
+					ClientID:           "baz",
+					UserPrefix:         "foo",
+					ValidationKey:      base.StringPointer("qux"),
+					Register:           true,
+					DiscoveryURI:       issuerFoo + auth.OIDCDiscoveryConfigPath,
+					IncludeAccessToken: true,
+				},
+			},
+			defaultProvider:      "foo",
+			wantUsername:         "foo_noah",
+			authURL:              "/db/_oidc?provider=foo&offline=true",
+			register:             true,
+			includeAccessToken:   true,
+			wantRefreshNoIDToken: true,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			opts := auth.OIDCOptions{Providers: tc.providers, DefaultProvider: &tc.defaultProvider}
+			restTesterConfig := RestTesterConfig{DatabaseConfig: &DbConfig{OIDCConfig: &opts}}
+			restTester := NewRestTester(t, &restTesterConfig)
+			defer restTester.Close()
+			mockSyncGateway := httptest.NewServer(restTester.TestPublicHandler())
+			defer mockSyncGateway.Close()
+			mockSyncGatewayURL := mockSyncGateway.URL
+
+			if tc.wantNoCode {
+				mockAuthServer.SetOption("wantNoCode", true)
+				defer mockAuthServer.DeleteOption("wantNoCode")
+			}
+			if tc.wantCallbackError {
+				mockAuthServer.SetOption("wantCallbackError", true)
+				defer mockAuthServer.DeleteOption("wantCallbackError")
+			}
+			if tc.wantUntoldProvider {
+				mockAuthServer.SetOption("wantUntoldProvider", true)
+				defer mockAuthServer.DeleteOption("wantUntoldProvider")
+			}
+			if tc.wantDiscoveryError {
+				mockAuthServer.SetOption("wantDiscoveryError", true)
+				defer mockAuthServer.DeleteOption("wantDiscoveryError")
+			}
+			if tc.wantCallbackTokenExgError {
+				mockAuthServer.SetOption("wantTokenExchangeError", true)
+				defer mockAuthServer.DeleteOption("wantTokenExchangeError")
+			}
+			if tc.wantCallbackNoIDToken {
+				mockAuthServer.SetOption("wantNoIDToken", true)
+				defer mockAuthServer.DeleteOption("wantNoIDToken")
+			}
+
+			// Initiate OpenID Connect Authorization Code flow.
+			requestURL := mockSyncGatewayURL + tc.authURL
+			request, err := http.NewRequest(http.MethodGet, requestURL, nil)
+			require.NoError(t, err, "Error creating new request")
+			response, err := http.DefaultClient.Do(request)
+			require.NoError(t, err, "Error sending request")
+
+			if !tc.register {
```

Comment:
As discussed, it's preferable to avoid a long list of conditions like this, when only a single test case is expected to evaluate to true.  In this case, since we're always evaluating the http response for the initial call, a test parameter for the error would be preferable (and would also make it obvious what each test is evaluating from the test case definition).